Proposal for Mimic Constraint Feature

scpeters opened the pull request:
This proposal suggests a new `mimic` constraint that is functionally similar to the `gearbox` joint type but simpler to express and more flexible. It is targeted at SDFormat 1.11 / libsdformat14.

The proposal is ready for review:

Preview: http://sdformat.org/tutorials?tut=mimic_proposal&cat=pose_semantics_docs&branch=scpeters%2Fmimic_joint_proposal

Related pull requests:

* [x] https://github.com/gazebosim/sdformat/pull/1166
* [ ] https://github.com/gazebosim/gz-physics/pull/431
  * [ ] https://github.com/gazebosim/gz-physics/issues/432
* [ ] https://github.com/gazebosim/gz-sim/pull/1838

scpeters commented:
cc @adityapande-1995

scpeters commented on mimic_joint/proposal.md:
```diff
@@ -0,0 +1,175 @@
+the following `mimic` joint:
+
+~~~
+        <joint name="mimic_demo" type="mimic">
```
actually, URDF has a `<mimic />` tag that can apply to any `<joint />` instead of a `type`; let me think about that

scpeters replied:
https://github.com/ros/urdfdom/blob/36be7f8f12c6269483b9f0d9c85b5a92a1c9e7ad/urdf_parser/test/urdf_unit_test.cpp#L184

scpeters replied:
I think I would be inclined to add `//joint/axis/mimic` and `//joint/axis2/mimic` tags instead of adding a new joint type. Let's update the proposal accordingly

adityapande-1995 replied:
Edit : https://github.com/gazebosim/sdf_tutorials/pull/80

scpeters commented:
I've updated the proposed syntax in https://github.com/gazebosim/sdf_tutorials/pull/62/commits/806841a8b49297e3636585e704e70866fc0fb2b8 and added a rack and pinion example in https://github.com/gazebosim/sdf_tutorials/pull/62/commits/c8443edfe11ab6b72238ac0ca4ee11ed866435d6

scpeters commented:
If a joint wants to mimic an axis of a multi-axis joint, like a universal joint, I wonder how we should specify that? Currently `//mimic/@joint` refers to a joint name, but we could say `joint::axis_index` like the following:

~~~
<mimic joint="joint_name::1"/>
~~~

in that case, perhaps the attribute should be `joint_axis`? or should we have two attributes? does the axis index start at 0 or 1?

~~~
<mimic joint="joint_name" axis="1"/>
~~~

adityapande-1995 commented:
Since the second tag is called ``<axis2>``, probably 1 and 2 makes sense. OR we can do ``joint_name::axis2``.

scpeters commented:
I just added a `//mimic/reference` parameter in https://github.com/gazebosim/sdf_tutorials/pull/62/commits/93a8fadd594a604a920584fe92acc8a007cb0fe2

adityapande-1995 commented:
> I just added a `//mimic/reference` parameter in [93a8fad](https://github.com/gazebosim/sdf_tutorials/commit/93a8fadd594a604a920584fe92acc8a007cb0fe2)

Added to implementation here : https://github.com/gazebosim/gz-physics/pull/431/commits/943740c85adc8d6266fc5a6818ade5b9985467d2

scpeters commented:
this is ready for review

azeey commented on mimic_constraint/proposal.md:
```diff
@@ -0,0 +1,318 @@
+~~~
+
+
+## Appendix
```
Remove

scpeters replied:
https://github.com/gazebosim/sdf_tutorials/pull/62/commits/57aaca17bc43cffbe62f16ab62ba9655f8c52723

azeey commented on mimic_constraint/proposal.md:
```diff
@@ -0,0 +1,318 @@
+Steve Peters`<scpeters@openrobotics.org>`,
+Aditya Pande `<aditya.pande@osrfoundation.org>`
+* **Status**: *Draft*
+* **SDFormat Version**: *1.10+*
```
Should this be 1.11?

scpeters replied:
we had been targeting garden / 1.10 since October 2021, but I can retarget to Harmonic / 1.11. I'll need to update the other pull requests; please confirm

scpeters replied:
changed to 1.11 in https://github.com/gazebosim/sdf_tutorials/pull/62/commits/57aaca17bc43cffbe62f16ab62ba9655f8c52723; I will proceed with retargeting the other pull requests

scpeters commented on mimic_constraint/proposal.md:
```diff
@@ -0,0 +1,318 @@
+        </joint>
+~~~
+
+The `gearbox` joint could replaced equivalently by adding the ``<mimic>``
```
https://github.com/gazebosim/sdf_tutorials/pull/62/commits/57aaca17bc43cffbe62f16ab62ba9655f8c52723
